### PR TITLE
Bump LLVM to v18

### PR DESCRIPTION
**Pull request by Zentrik**

### CODE REVIEW

File: src/disasm.cpp [hunk 1]
```diff
@@ -1224,7 +1224,11 @@ jl_value_t *jl_dump_function_asm_impl(jl_llvmf_dump_t* dump, char emit_mc, const
         addTargetPasses(&PM, TM->getTargetTriple(), TM->getTargetIRAnalysis());
         if (emit_mc) {
             raw_svector_ostream obj_OS(ObjBufferSV);
-            if (TM->addPassesToEmitFile(PM, obj_OS, nullptr, CGFT_ObjectFile, false, nullptr))
+#if JL_LLVM_VERSION >= 180000
+            if (TM->addPassesToEmitFile(PM, obj_OS, nullptr, CodeGenFileType::ObjectFile, false, nullptr))
+#else
+            if (TM->addPassesToEmitFile(PM, obj_OS, nullptr, CodeGenFileType::ObjectFile, false, nullptr))
```
Comment (Zentrik) [inline src/disasm.cpp:1230]:
```suggestion
            if (TM->addPassesToEmitFile(PM, obj_OS, nullptr, CGFT_ObjectFile, false, nullptr))
```

**Comment by Zentrik** [comment]
Bit confused why windows won't start building

**Comment by giordano** [comment]
I think there's something broken with the CI setup. CC: @staticfloat

File: src/datatype.c [hunk 2]
```diff
@@ -935,6 +935,10 @@ JL_DLLEXPORT jl_datatype_t *jl_new_primitivetype(jl_value_t *name, jl_module_t *
                                         jl_emptysvec, jl_emptysvec, jl_emptysvec, 0, 0, 0);
     uint32_t nbytes = (nbits + 7) / 8;
     uint32_t alignm = next_power_of_two(nbytes);
+# if defined(_CPU_X86_) && !defined(_OS_WINDOWS_)
+    if (alignm == 8)
```
Comment (gbaraldi) [inline src/datatype.c:939]: Is this correct? https://godbolt.org/z/ec3bbo69T seems to show that it's 8 byte aligned

Comment (Zentrik) [reply src/datatype.c:939]: The godbolt link is for windows right? and this code is for not windows.

Comment (gbaraldi) [reply src/datatype.c:939]: Oh I missed the `!`, you're right

**Comment by gbaraldi** [comment]
This fixes the llvmpasses tests
```diff
diff --git a/test/llvmpasses/pipeline-prints.ll b/test/llvmpasses/pipeline-prints.ll
index babd26c797..ecb7095302 100644
--- a/test/llvmpasses/pipeline-prints.ll
+++ b/test/llvmpasses/pipeline-prints.ll
@@ -298,12 +298,12 @@ attributes #2 = { inaccessiblemem_or_argmemonly }

 ; COM: Loop simplification makes the exit condition obvious
 ; AFTERLOOPSIMPLIFICATION: L35.lr.ph:
-; AFTERLOOPSIMPLIFICATION-NEXT: add nuw nsw
+; AFTERLOOPSIMPLIFICATION: add nuw nsw

 ; COM: Scalar optimization removes the previous add from the preheader
-; AFTERSCALAROPTIMIZATION: L35.preheader:
+; AFTERSCALAROPTIMIZATION: L35.lr.ph:
 ; AFTERSCALAROPTIMIZATION-NOT: add nuw nsw
-; AFTERSCALAROPTIMIZATION-NEXT: br label %L35
+; AFTERSCALAROPTIMIZATION: br label %L35

 ; COM: Vectorization does stuff
 ; AFTERVECTORIZATION: vector.body
```

**Comment by vtjnash** [comment]
https://github.com/llvm/llvm-project/pull/78251

**Comment by gbaraldi** [comment]
The windows build is segfaulting in 
> #0  0x00007ff8e4f70de5 in ntdll!RtlVirtualUnwind () from /c/Windows/SYSTEM32/ntdll.dll
#1  0x00007ff8e4f70ccb in ntdll!RtlVirtualUnwind () from /c/Windows/SYSTEM32/ntdll.dll
#2  0x00007ff8e3ee3bd8 in RtlVirtualUnwind () from /c/Windows/System32/KERNEL32.DLL
#3  0x00007ff8bb29cfed in jl_unw_step (cursor=0xba5690, from_signal_handler=0, ip=<synthetic pointer>,
    sp=<synthetic pointer>) at C:/msys64/home/vboxuser/julia/src/stackwalk.c:523
#4  jl_unw_stepn (cursor=cursor@entry=0xba5690, bt_data=bt_data@entry=0x42b0080, bt_size=bt_size@entry=0xba51b8,
    sp=sp@entry=0x0, maxsize=maxsize@entry=80000, skip=0, skip@entry=3, ppgcstack=ppgcstack@entry=0xba51b0,
    from_signal_handler=from_signal_handler@entry=0) at C:/msys64/home/vboxuser/julia/src/stackwalk.c:99
#5  0x00007ff8bb29d752 in rec_backtrace (bt_data=0x42b0080, maxsize=maxsize@entry=80000, skip=skip@entry=2)
    at C:/msys64/home/vboxuser/julia/src/stackwalk.c:222
#6  0x00007ff8bb264313 in record_backtrace (ptls=0x3d71d10, skip=skip@entry=1)
    at C:/msys64/home/vboxuser/julia/src/task.c:412
#7  0x00007ff8bb264a85 in ijl_throw (e=0x13b7c0c0) at C:/msys64/home/vboxuser/julia/src/task.c:771
#8  0x00007ff8bb2cfdb9 in ijl_errorf (
    fmt=fmt@entry=0x7ff8bb482178 <flisp_system_image+362104> "Type %s does not have a definite size.")
    at C:/msys64/home/vboxuser/julia/src/rtutils.c:77
#9  0x00007ff8bb24d3a8 in jl_f_sizeof (F=<optimized out>, args=<optimized out>, nargs=<optimized out>)
    at C:/msys64/home/vboxuser/julia/src/builtins.c:531

**Comment by giordano** [comment]
Was libllvm_jll updated to 18.1.7+2?  That includes https://github.com/llvm/llvm-project/pull/88172 which should fix WIndows issues according to https://github.com/JuliaLang/llvm-project/pull/30#issue-2382476281.

**Comment by gbaraldi** [comment]
This doesn't have the new LLVM build yet. I was fixing the embedding thing mostly

**Comment by Zentrik** [comment]
Oops, 64 bit windows builds on https://github.com/Zentrik/llvm-project/commit/b557165510547af5b6806e2b3542e5210954645d not with https://github.com/JuliaLang/llvm-project/pull/30.

**Comment by gbaraldi** [comment]
Do we know which one broke it?

**Comment by Zentrik** [comment]
I have no clue.

**Comment by Zentrik** [comment]
Looks like https://github.com/Zentrik/llvm-project/commit/5d8cd9873ec596c7a7cae56b476e5339f171e6be is sufficient to get llvm 18 building on 64 bit windows. The pr introducing the change, https://github.com/llvm/llvm-project/pull/73037 suggests that the pr could cause trouble if we are assuming/hardcoding '.text'.

**Comment by gbaraldi** [comment]
I'm tempted to remove that assertion. It's triggering in quite a few places and doesn't seem to guard anything that matters.

**Comment by gbaraldi** [comment]
What is segfaulting here? Because I don't see how that breaks stuff. We just register functions with the JIT

**Comment by Zentrik** [comment]
This assertion triggers on a .ltext section https://github.com/JuliaLang/julia/blob/0ef2bb61e0900319cf22b2b2c15d0bdb2cdfa5c3/src/cgmemmgr.cpp#L868

**Comment by gbaraldi** [comment]
But if we remove the assertion (it seems slightly weird anyway)

**Comment by Zentrik** [comment]
I didn't look into the segfaults at all so no idea. Pretty sure removing the assertion will just lead to another one being triggered, I recall at some point having https://github.com/llvm/llvm-project/blob/73447a3302541c343525570222b318e7f94f9402/llvm/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp#L299-L301 assert though I don't remember how I got to that point.

**Comment by Zentrik** [comment]
Commenting out https://github.com/JuliaLang/julia/blob/04af4460e6863c5bf4e5c5e01a389d5fb39e56d4/src/debuginfo.cpp#L355-L364 and https://github.com/JuliaLang/julia/blob/04af4460e6863c5bf4e5c5e01a389d5fb39e56d4/src/cgmemmgr.cpp#L868 gets windows 64 bit building with assertions.



When building without commenting out the above two sections, first https://github.com/JuliaLang/julia/blob/04af4460e6863c5bf4e5c5e01a389d5fb39e56d4/src/cgmemmgr.cpp#L868 triggers after `allocateCodeSection` is called to allocate a `.text` and then a `.ltext` without a call to `finalizeMemory`. Whereas with https://github.com/Zentrik/llvm-project/commit/5d8cd9873ec596c7a7cae56b476e5339f171e6be there is only a call to allocate a `.text`. Disabling this assertion then causes https://github.com/JuliaLang/julia/blob/04af4460e6863c5bf4e5c5e01a389d5fb39e56d4/src/debuginfo.cpp#L357-L358 to fail on a symbol in '.ltext'. Disabling this as well, causes https://github.com/JuliaLang/julia/blob/04af4460e6863c5bf4e5c5e01a389d5fb39e56d4/src/debuginfo.cpp#L182 to fail on the same symbol.

**Comment by Zentrik** [comment]
I suspect https://github.com/JuliaLang/julia/blob/04af4460e6863c5bf4e5c5e01a389d5fb39e56d4/src/jitlayers.cpp#L2242-L2255 might be causing problems but not sure.

**Comment by Zentrik** [comment]
I think using a linker script like https://lore.kernel.org/linux-kernel/46343CD5-24D6-46BF-92A7-0B0FA0E6937D@alien8.de/T/#Z2e.:..:20240417-x86-fix-kexec-with-llvm-18-v1-1-5383121e8fb7::40kernel.org:1arch:x86:purgatory:.gitignore would work but seems hacky. 

Alternatively, I think if the assembly above was added to .ltext that should fix things, but I can't figure out how to.

**Comment by Zentrik** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by Zentrik** [comment]
@nanosoldier `runtests()`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/54848#issuecomment-2241635304) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/8094ce4_vs_680c7e3/report.md).

**Comment by Zentrik** [comment]
Benchmarks lgtm assuming the regressions in scalar and collections on the order of ns are noise. An average .5% perf improvement excluding those.

**Comment by giordano** [comment]
> assuming the regressions in scalar and collections on the order of ns are noise.

I ran locally some benchmarks on `sincos` and `sinh` (among the scalar functions with worse regressions) and I couldn't see any sensible difference between this PR
```
julia> versioninfo()
Julia Version 1.12.0-DEV.899
Commit b9281e34aa8 (2024-07-21 16:17 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  WORD_SIZE: 64
  LLVM: libLLVM-18.1.7 (ORCJIT, haswell)
Threads: 1 default, 0 interactive, 1 GC (on 8 virtual cores)
```
and latest nightly
```
julia> versioninfo()
Julia Version 1.12.0-DEV.882
Commit a1e0f5d921c (2024-07-22 01:06 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
  WORD_SIZE: 64
  LLVM: libLLVM-17.0.6 (ORCJIT, haswell)
Threads: 1 default, 0 interactive, 1 GC (on 8 virtual cores)
```

**Comment by oscardssmith** [comment]
is the generated code similar?

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/54848#issuecomment-2241709430) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/67794fc_vs_680c7e3/report.html) is available.

**Comment by Zentrik** [comment]
I suspect the two packages failing due to invalid IR generated is due to https://github.com/llvm/llvm-project/issues/63984.

**Comment by gbaraldi** [comment]
The `skipmissing` regression is real if anyone wants a deeper look. I really wish BaseBenchmarks had an option to ask for the LLVM IR of the function being called.

**Comment by Zentrik** [comment]
> The `skipmissing` regression is real if anyone wants a deeper look. I really wish BaseBenchmarks had an option to ask for the LLVM IR of the function being called. 😉

The unoptimized LLVM IR looks the same to me, here's the optimized IR with LLVM 18 https://gist.github.com/Zentrik/b49a9d5b67dcfaae7b420ef8ed28df03 and on 0fd1f04dc7d https://gist.github.com/Zentrik/e42698cb8db60932790f4f3657d35efb. The main difference seems to be a error branch isn't optimised out.

**Comment by gbaraldi** [comment]
I think this is ready to merge. @vtjnash and @vchuravy any objections?

**Comment by vchuravy** [comment]
Congratulations! Awesome work everyone :)

**Pull request merged**